plugins/user_mapping/src/local_file.cpp

```diff
@@ -0,0 +1,173 @@
+#include "interface.h"
+
+#include <cstdlib>
+#include <cstring>
+#include <exception>
+#include <filesystem>
+#include <fstream>
+#include <mutex>
+#include <optional>
+#include <shared_mutex>
+#include <string>
+#include <vector>
+
+#include <nlohmann/json.hpp>
+#include <spdlog/spdlog.h>
+
+namespace
+{
+	struct user_profile
+	{
+		std::string irods_user_name;
+		nlohmann::json attributes;
+	};
+
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+	std::string file_path;
+
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+	std::vector<user_profile> profile_list;
+
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables, cert-err58-cpp)
+	std::shared_mutex list_mutex;
+
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+	std::mutex update_mutex;
+
+	auto init(const nlohmann::json& _config) -> void
+	{
+		const auto path{_config.find("file_path")};
+		if (path == std::end(_config)) {
+			// handle bad case
+			spdlog::error("{}: Unable to find [file_path] in provided config.", __func__);
+		}
+
+		// Save file string
+		file_path = path->get<std::string>();
+	} // init
+
+	auto update() -> void
+	{
+		static std::filesystem::file_time_type last_file_path_write;
+
+		// If there have been no changes to file_path, there is no work to do
+		if (std::filesystem::last_write_time(file_path) == last_file_path_write) {
+			return;
+		}
+
+		std::ifstream file{file_path};
+		spdlog::debug("{}: Opened file [{}]", __func__, file_path);
```

Comment:
fstreams never throw (unless configured to).

You need to test the stream to make sure it was opened successfully.
```cpp
if (!file) {
    // Could not open file for reading. Handle error.
}
```

Reply:
Same question as this: https://github.com/irods/irods_client_http_api/pull/331#discussion_r1774042935

Do we throw or return here, since there's an exception handler up the stack?

Reply:
Added `if` statement with log, no return or throw yet.

Reply:
Throwing is fine as long as we let the exception escape the plugin.

Reply:
'don't' ?

Reply:
Sorry. I meant "don't".

Throwing is fine as long as we **don't** let exceptions cross the plugin boundary.

Reply:
Throws now. Should we still log if the thrown message is the same?

Reply:
See https://github.com/irods/irods_client_http_api/pull/331#discussion_r1775456139

Reply:
Done.